test/test_aggregate_functions.js

```diff
@@ -0,0 +1,106 @@
+exports.test = function (SQL, assert) {
+    function assertFloat(got, expected, message="", sigma=0.001) {
+        assert.ok(got > expected - sigma && got < expected + sigma, message);
+    }
+
+    var db = new SQL.Database();
+
+    db.create_aggregate(
+        "sum", 0, {
+            step: function (state, value) { return state + value; },
+        }
+    );
+
+    db.exec("CREATE TABLE test (col);");
+    db.exec("INSERT INTO test VALUES (1), (2), (3), (null);");
+    var result = db.exec("SELECT sum(col) FROM test;");
+    assert.equal(result[0].values[0][0], 6, "Simple aggregate function.");
+
+    db.create_aggregate(
+        "percentile", { vals: [], pctile: null },
+        {
+            step: function (state, value, pctile) {
+                var typ = typeof value;
+                if (typ == "number" || typ == "bigint") {
+                    state.pctile = pctile;
+                    state.vals.push(value);
+                }
+                return state;
+            },
+            finalize: function (state) {
+                return percentile(state.vals, state.pctile);
+            }
+        }
+    );
+    result = db.exec("SELECT percentile(col, 80) FROM test;");
+    assertFloat(result[0].values[0][0], 2.6, "Aggregate function with two args");
+
+    db.create_aggregate(
+        "json_agg", [], {
+            step: (state, val) => [...state, val],
+            finalize: (state) => JSON.stringify(state),
+        }
+    );
+
+    db.exec("CREATE TABLE test2 (col, col2);");
+    db.exec("INSERT INTO test2 values ('four score', 12), ('and seven', 7), ('years ago', 1);");
+    result = db.exec("SELECT json_agg(col) FROM test2;");
+    assert.deepEqual(
+        JSON.parse(result[0].values[0]),
+        ["four score", "and seven", "years ago"],
+        "Aggregate function that returns JSON"
+    );
+
+    result = db.exec("SELECT json_agg(col), json_agg(col2) FROM test2;");
+    assert.deepEqual(
+        result[0].values[0].map(JSON.parse),
+        [["four score", "and seven", "years ago"], [12, 7, 1]],
+        "Multiple aggregations at once"
+    );
+
+    db.create_aggregate("throws_step", 0, {step: (state, value) => { throw "bananas" }})
+    assert.throws(
+        () => db.exec("SELECT throws_step(col) FROM test;"),
+        "Error: bananas",
+        "Handles exception in a step function"
+    );
```

Comment:
Maybe the test here should test more things ? The exception should not have messed the internal state, and the function should still work after the throw. 

Maybe something like : 

```suggestion
    db.create_aggregate("sum_non_zero", {init: () => 0, step: (state, value) => { 
        if (value==0) throw "bananas";
        return state + value  
    }})
    assert.throws(
        () => db.exec("SELECT sum_non_zero(column1) FROM (VALUES (1),(2),(0))"),
        "Error: bananas",
        "Handles exception in a step function"
    );
    assert.deepEqual(
        db.exec("SELECT sum_non_zero(column1) FROM (VALUES (1),(2))")[0].values[0][0],
        [3],
        "1 + 2 = 3"
    );
```

Reply:
added for both exception tests